File: questions/227. Basic Calculator II/basic.cpp
class Solution {
public:
    int calculate(string s) {
        int cur_num = 0, res = 0;
        char sign = '+';
        stack<int> st;
        
        for(int i = 0; i < s.size(); i++){
            char c = s[i];
            if(c >= '0' && c <= '9'){
                cur_num = cur_num*10 + (int)(c-'0');
            }
            if(i == s.size() - 1 || (!(c >= '0' && c <= '9') && ' ' != c)){
                if(sign == '+'){
                    st.push(cur_num);
                }else if(sign == '-'){
                    st.push(-cur_num);
                }else if(sign == '*'){
                    int n = st.top();
                    st.pop();
                    st.push(n * cur_num);
                }else if(sign == '/'){
                    int n = st.top();
                    st.pop();
                    st.push(n/cur_num);
                }
                sign = c;
                cur_num = 0;
            }
        }
        while(!st.empty()){
            res += st.top();
            st.pop();
        }
        return res;
    }
};

File: questions/227. Basic Calculator II/basic.java
class Solution {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack();
        int num = 0;
        char sign = '+';
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(Character.isDigit(c)){
                num = num * 10 + (int)(c - '0');
            }
            if(i == s.length() - 1 || (!Character.isDigit(c) && c != ' ')){
                if(sign == '+'){
                    stack.push(num);
                }else if(sign == '-'){
                    stack.push(-num);
                }else if(sign == '*'){
                    stack.push(num*stack.pop());
                }else if(sign == '/'){
                    stack.push(stack.pop()/num);
                }
                sign = c;
                num = 0;
            }
        }
        for(int i : stack){
            num += i;
        }
        return num;
    }
}
